stablehlo/dialect/TypeInference.cpp

```diff
@@ -1604,33 +1604,37 @@ LogicalResult inferComplexOp(std::optional<Location> location, Value lhs,
 LogicalResult inferConcatenateOp(std::optional<Location> location,
                                  TypeRange inputTypes, int64_t dimension,
                                  SmallVectorImpl<Type>& inferredReturnTypes) {
+  // concatenate_c4
   if (dimension < 0)
     return emitOptionalError(location, "dimension ", dimension, " is negative");
   RankedTensorType firstRankedType;
   int firstRankedIndex = -1;
   for (uint64_t i = 0; i < inputTypes.size(); i++) {
     auto secondType = inputTypes[i].dyn_cast<ShapedType>();
     if (!secondType.hasRank()) continue;
-
     if (!firstRankedType) {
       firstRankedType = secondType.cast<RankedTensorType>();
       firstRankedIndex = i;
+      // concatenate_c4
       if (firstRankedType.getRank() == 0)
         return emitOptionalError(location,
                                  "rank-0 values cannot be concatenated");
+      // concatenate_c4
       if (dimension >= firstRankedType.getRank())
         return emitOptionalError(location, "dimension ", dimension,
                                  " is out-of-bounds for input rank ",
                                  firstRankedType.getRank());
       continue;
     }
+    // concatenate_c2
     if (firstRankedType.getRank() != secondType.getRank())
       return emitOptionalError(location, "operands (", firstRankedIndex,
                                ") and (", i, ") do not match rank");
 
     auto firstShape = firstRankedType.getShape();
     auto secondShape = secondType.getShape();
     for (int d = 0; d < firstRankedType.getRank(); ++d) {
+      // concatenate_c2, concatenate_c6
```

Comment:
Why is `concatenate_c6` involved here? It's talking about the shape of the result, right?

Reply:
Agreed.